client/jetpack-cloud/sections/pricing/jpcom-masterbar/components/user-menu.tsx

```diff
@@ -0,0 +1,87 @@
+import { localizeUrl, useLocale } from '@automattic/i18n-utils';
+import classNames from 'classnames';
+import { useTranslate } from 'i18n-calypso';
+import { useState, useCallback } from 'react';
+import Gravatar from 'calypso/components/gravatar';
+import { useSelector } from 'calypso/state';
+import { isUserLoggedIn, getCurrentUser } from 'calypso/state/current-user/selectors';
+import type { MouseEvent as ReactMouseEvent, FC } from 'react';
+
+const UserMenu: FC = () => {
+	const locale = useLocale();
+	const isLoggedIn = useSelector( isUserLoggedIn );
+	const user = useSelector( getCurrentUser );
+	const translate = useTranslate();
+	const [ isMenuOpen, setIsMenuOpen ] = useState( false );
+
+	const onDocumentClick = () => {
+		setIsMenuOpen( false );
+	};
+
+	const toggleUserMenu = useCallback( () => {
+		if ( isMenuOpen ) {
+			document.removeEventListener( 'click', onDocumentClick );
+		} else {
+			document.addEventListener( 'click', onDocumentClick );
+		}
+
+		setIsMenuOpen( ( open ) => ! open );
+	}, [ isMenuOpen ] );
+
+	const onUserBtnClick = ( e: ReactMouseEvent< HTMLAnchorElement, MouseEvent > ) => {
+		e.preventDefault();
+		e.stopPropagation();
+
+		toggleUserMenu();
+	};
+
+	return (
+		<ul className="header__actions-list">
+			<li
+				className={ classNames( 'header__user-menu user-menu ', {
+					'is-logged-in': isLoggedIn,
+				} ) }
+			>
+				{ isLoggedIn ? (
+					<>
```

Comment:
I think we can limit such deep hierarchies of react components by defining some local view components.